gradle-jdks/src/main/java/com/palantir/gradle/jdks/flow/ToolchainFlowAction.java

```diff
@@ -0,0 +1,97 @@
+ */
+
+package com.palantir.gradle.jdks.flow;
+
+import com.google.common.base.Throwables;
+import com.palantir.gradle.jdks.flow.ToolchainFlowAction.Parameters;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.gradle.api.flow.BuildWorkResult;
+import org.gradle.api.flow.FlowAction;
+import org.gradle.api.flow.FlowParameters;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class ToolchainFlowAction implements FlowAction<Parameters> {
+
+    private static final Logger log = LoggerFactory.getLogger(ToolchainFlowAction.class);
+
+    interface Parameters extends FlowParameters {
+        @Input
+        Property<BuildWorkResult> getBuildResult();
+
+        @Input
+        ListProperty<String> getConfiguredJavaMajorVersions();
+    }
+
+    @Override
+    @SuppressWarnings("LineLength")
+    public void execute(Parameters parameters) {
+        parameters.getBuildResult().get().getFailure().ifPresent(failure -> {
+            List<Throwable> noToolchainsAvailable = Throwables.getCausalChain(failure).stream()
+                    .filter(throwable ->
+                            throwable instanceof org.gradle.jvm.toolchain.internal.NoToolchainAvailableException)
+                    .collect(Collectors.toList());
+            if (noToolchainsAvailable.isEmpty()) {
+                return;
+            }
+            List<String> missingToolchains = noToolchainsAvailable.stream()
+                    .map(exception -> {
+                        Pattern pattern = Pattern.compile("languageVersion=(\\d+)");
+                        Matcher matcher = pattern.matcher(exception.getMessage());
+                        if (matcher.find()) {
+                            return Optional.of(matcher.group(1));
+                        }
+                        return Optional.<String>empty();
+                    })
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toList());
+            String maybeMissingToolchains = missingToolchains.isEmpty()
+                    ? "some java versions"
+                    : String.format("the java versions=%s", missingToolchains);
+            String includeVersionsOption = missingToolchains.isEmpty()
+                    ? "-includeVersion=<newJavaMajorVersion>"
+                    : missingToolchains.stream()
+                            .map(version -> String.format("--includeVersion=%s", version))
+                            .collect(Collectors.joining(" "));
+            log.error(
+                    "\n"
+                        + "\u001B[31m****************************************************************************************************\n"
```

Comment:
Do we need `\u001B[31m` to change to red? I thought error logs were red in gradle at least?

fwiw I think if used on circle the `\u001B[31m` will look weird as it's not a terminal

Reply:
I tried now without the `\u001B[31m` and it doesn't change it to red :(
<img width="1570" alt="Screenshot 2025-04-07 at 3 50 39 PM" src="https://github.com/user-attachments/assets/53472228-2ca4-42f4-bb0a-776c7922576f" />

Reply:
Ah, I'd at least only output the control characters when not on CI then.